### ISSUE TITLE
fastest way to predict structures for a large library of sequences

### ISSUE THREAD

**Issue by sahakyanhk** [ml4bio/RhoFold#14]
Is it possible to predict structures for many different RNA sequences with a single pass? 

I have a large database of sequences and want to predict their structures without building MSAs. What would be the fastest way? Could you provide any code snippets for such cases? 

Thanks!

**Comment by WangJiuming** [comment]
Hi @sahakyanhk,

Thanks a lot for this feature request. If the RNA sequences are kept in individual `.fasta` files, then a bash script like below may help:

```bash
fasta_dir="path/to/fasta/files"
output_root_dir="path/to/output/dirs"

for fasta_path in $fasta_dir/*.fasta; do

    data_name=$(basename "${fasta_path%.fasta}")
    echo $data_name

    input_fasta_path=$fasta_path
    output_dir="$output_root_dir/$data_name"
    mkdir -p $output_dir

    python rhofold/inference.py --input_fas "${input_fasta_path}" \
                                --output_dir "${output_dir}" \
                                --device "${device}" \
                                --single_seq_pred
done
```

You can set up the `fasta_dir` and `output_root_dir` variables in the beginning.

Feel free to let us know if there is any follow-up question.

**Comment by sahakyanhk** [comment]
Hi, thanks for the code, but I mean running only the inference without the entire pipeline. 

Let's say something like this 

```
@torch.no_grad()
def rnafold(data_dict):
    outputs = model(tokens=data_dict['tokens'].to(device),
                rna_fm_tokens=data_dict['rna_fm_tokens'].to(device),
                seq=data_dict['seq'],
                )

    return outputs
```

where `data_dict` contains information for several sequences.

**Comment by WangJiuming** [comment]
Hi @sahakyanhk,

I suppose you have one `fasta` file containing multiple sequences and would like to obtain the results without using MSA.

To do that, I have added a new script `rhofold/batch_inference.py` at https://github.com/WangJiuming/rhofold_protocol. You can run it as follows:

```bash
python rhofold/batch_inference.py --fasta path/to/your/fasta --output-dir path/to/your/output/root/dir --relax-steps 0 --device cuda:0
```

This will allow inputting a single `fasta` file containing all sequences, then process and save them with one command.

Note that this script assumes no MSA input, and the above command does not perform any relaxation (`--relax-steps 0`) in order to save time. Only the structure prediction step is performed.

**Comment by sahakyanhk** [comment]
Hi @WangJiuming, thanks for the new script, but as I understand, `rhofold/batch_inference.py` runs inference within a for loop, one sequence at a time. I thought there might be an option to predict a batch of sequences at once. For instance, as it was implemented in ESMfold, where a list of sequences can be inferred. https://github.com/facebookresearch/esm/blob/main/esm/esmfold/v1/esmfold.py#L281

**Comment by ATP-BME** [comment]
Hi there,
I am also facing the exact same problem.I would like to perform real native batch inference, instead of simply running single-sequence inference in a for-loop wrapper.
In my use case, the sequences in one batch have different lengths. Could you kindly share how the model handles batch input during training?Specifically, how does the training pipeline deal with variable-length sequences within the same batch?
It would be very helpful if we could adopt the same batch padding / masking strategy used in training for inference, so that we can run efficient true batch inference for multi-length protein sequences.
Thank you very much for your time and support!

**Comment by WangJiuming** [comment]
Hi @ATP-BME and @sahakyanhk, 

I've added a new branch for batch inference at `https://github.com/WangJiuming/rhofold_protocol/tree/dev/batch_inference`. The new codes use length-bucketing and padding for batching the sequences. Feel free to let me know at the new protocol repo as well if there is any follow-up question.
